design/multiple-listeners-design.md

```diff
@@ -0,0 +1,147 @@
+In Gateway mode, Contour is configured to process a Gateway API Gateway, and gets its Listener configuration from there.
+Non-Gateway mode is the traditional mode of operation, where Listener configuration comes from CLI flags, a config file, and/or a ContourConfiguration resource, plus the Envoy service YAML, and HTTPProxy and/or Ingress is used for routing configuration.
+
+For non-Gateway mode, this design should be fully backwards-compatible.
+There should be no changes observed by a user of Contour in non-Gateway mode.
+
+For Gateway mode, the following changes are needed:
+
+- The Gateway provisioner must configure ports on the Envoy service and daemonset/deployment corresponding to all of the Gateway Listener ports. The mapping between service ports and container ports will be described in the Detailed Design section below.
+- Contour's directed acyclic graph (DAG) processing changes, to first add Listeners, and then add virtual hosts directly to the appropriate Listener(s).
+- Each DAG Listener results in an Envoy Listener being configured. The naming scheme will be described in the Detailed Design section below. 
+- Envoy RouteConfigurations will be generated for each Envoy Listener. The naming scheme will be described in the Detailed Design section below. 
+
+For Gateway mode, we must also consider what happens when HTTPProxy or Ingress is used for routing configuration.
+The details of this use case are described in the Compatibility section below.
+
+## Detailed Design
+
+### internal/gatewayapi
+The `ValidateListeners` function logic must change, to allow any number of unique ports, while still validating protocols and hostnames for each port.
+Instead of returning a single insecure and secure port number, it will return a slice of ports.
+
+### internal/provisioner
+The Gateway controller will use the results of the `ValidateListeners` function (above) to add all relevant ports to `contourModel.Spec.NetworkPublishing.Envoy.Ports`.
+These ports will be populated on the Envoy service and deployment/daemonset.
+
+Currently, the single insecure port is always mapped to port 8080 inside the Envoy container, and the single secure port is always mapped to 8443.
+Supporting any number of Listeners requires a more flexible port mapping scheme, since the requested Gateway Listener ports could be privileged ports (e.g. 1-1023), non-privileged ports (e.g. 8080, 8443), and/or node ports (e.g. 32080).
+The main constraint is that, since Envoy runs as non-root, it cannot bind to privileged ports within the container.
+As such, any requested privileged Gateway Listener port must be mapped to a non-privileged port inside the Envoy container.
+The scheme proposed is as follows:
+- any privileged Gateway Listener port (1-1023) is mapped to the range 64513-65535 within the container.
+- any other Gateway Listener port (1024-65535) is mapped as-is within the container.
```

Comment:
Something we have to think about within this range is the admin and stats Listeners that Contour always creates that take up ports 9001 and 8000(?)